diff --git a/hostapd/config_file.c b/hostapd/config_file.c
index 0259f6b82..f579530b7 100644
--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -5241,6 +5241,12 @@ int hostapd_config_s1g_val(struct hostapd_config *conf)
 	int chan_width;
 	int errors = 0;
 
+	/* Special region is set for s1g. Reset from backup for a valid s1g to ht conversion */
+	if(conf->country[0] == 'Z' && conf->country[1] == 'Z') {
+		conf->country[0] = conf->op_country[0];
+		conf->country[1] = conf->op_country[1];
+	}
+
 	morse_set_s1g_ht_chan_pairs(conf->country);
 
 	if (conf->acs) {
diff --git a/hostapd/main.c b/hostapd/main.c
index c8ccd6aff..7d01b176a 100644
--- a/hostapd/main.c
+++ b/hostapd/main.c
@@ -33,6 +33,7 @@
 #include "eap_register.h"
 #include "ctrl_iface.h"
 #include "build_features.h"
+#include "ap/ubus.h"
 
 struct hapd_global {
 	void **drv_priv;
@@ -1031,6 +1032,7 @@ int main(int argc, char *argv[])
 	}
 
 	hostapd_global_ctrl_iface_init(&interfaces);
+	hostapd_s1g_ubus_add_main(&interfaces);
 
 	if (hostapd_global_run(&interfaces, daemonize, pid_file)) {
 		wpa_printf(MSG_ERROR, "Failed to start eloop");
@@ -1041,6 +1043,8 @@ int main(int argc, char *argv[])
 
  out:
 	hostapd_global_ctrl_iface_deinit(&interfaces);
+	hostapd_s1g_ubus_free_main();
+
 	/* Deinitialize all interfaces */
 	for (i = 0; i < interfaces.count; i++) {
 		if (!interfaces.iface[i])
diff --git a/wpa_supplicant/wpa_supplicant.c b/wpa_supplicant/wpa_supplicant.c
index eb3296292..60538b411 100644
--- a/wpa_supplicant/wpa_supplicant.c
+++ b/wpa_supplicant/wpa_supplicant.c
@@ -1168,6 +1168,7 @@ void wpa_supplicant_set_state(struct wpa_supplicant *wpa_s,
 		sme_sched_obss_scan(wpa_s, 0);
 	}
 	wpa_s->wpa_state = state;
+	wpas_ubus_event_state(wpa_s, wpa_supplicant_state_txt(state));
 
 #ifdef CONFIG_BGSCAN
 	if (state == WPA_COMPLETED && wpa_s->current_ssid != wpa_s->bgscan_ssid)
