From 7e40114591cc8f8dec234cd185edc70ba6af02d1 Mon Sep 17 00:00:00 2001
From: Lyall Beveridge <lyall.beveridge@morsemicro.com>
Date: Tue, 14 Oct 2025 11:57:13 +1100
Subject: [PATCH] allow 0 BSS color

---
 src/utils/morse_cli.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/src/utils/morse_cli.c b/src/utils/morse_cli.c
index 83e211097..a317a76db 100644
--- a/src/utils/morse_cli.c
+++ b/src/utils/morse_cli.c
@@ -326,9 +326,6 @@ int morse_set_bss_color(const char *ifname, u8 s1g_bss_color)
 {
 	int ret;
 
-	if (!s1g_bss_color)
-		return 0;
-
 	ret = morse_cli(ifname, "bsscolor %d", s1g_bss_color);
 	if (ret != 0)
 		wpa_printf(MSG_WARNING,
-- 
2.43.0

