commit d40e6fbe047a4d207b354f29a4da7a808aea41c9
Author: Prabha Murugesan <prabha.murugesan@morsemicro.com>
Date:   Tue Mar 31 18:47:31 2026 +1100

    DPP: log RSSI for received action frames

--- a/src/ap/dpp_hostapd.c
+++ b/src/ap/dpp_hostapd.c
@@ -3058,7 +3058,8 @@
 
 
 void hostapd_dpp_rx_action(struct hostapd_data *hapd, const u8 *src,
-			   const u8 *buf, size_t len, unsigned int freq)
+			   const u8 *buf, size_t len, unsigned int freq,
+			   int ssi_signal)
 {
 	u8 crypto_suite;
 	enum dpp_public_action_frame_type type;
@@ -3082,25 +3083,28 @@
 
 	wpa_printf(MSG_DEBUG,
 		   "DPP: Received DPP Public Action frame crypto suite %u type %d from "
-		   MACSTR " freq=%u",
-		   crypto_suite, type, MAC2STR(src), freq);
+		   MACSTR " freq=%u rssi=%d",
+		   crypto_suite, type, MAC2STR(src), freq, ssi_signal);
 	if (crypto_suite != 1) {
 		wpa_printf(MSG_DEBUG, "DPP: Unsupported crypto suite %u",
 			   crypto_suite);
 		wpa_msg(hapd->msg_ctx, MSG_INFO, DPP_EVENT_RX "src=" MACSTR
-			" freq=%u type=%d ignore=unsupported-crypto-suite",
-			MAC2STR(src), freq, type);
+			" freq=%u type=%d ignore=unsupported-crypto-suite"
+			" rssi=%d",
+			MAC2STR(src), freq, type, ssi_signal);
 		return;
 	}
 	wpa_hexdump(MSG_MSGDUMP, "DPP: Received message attributes", buf, len);
 	if (dpp_check_attrs(buf, len) < 0) {
 		wpa_msg(hapd->msg_ctx, MSG_INFO, DPP_EVENT_RX "src=" MACSTR
-			" freq=%u type=%d ignore=invalid-attributes",
-			MAC2STR(src), freq, type);
+			" freq=%u type=%d ignore=invalid-attributes"
+			" rssi=%d",
+			MAC2STR(src), freq, type, ssi_signal);
 		return;
 	}
 	wpa_msg(hapd->msg_ctx, MSG_INFO, DPP_EVENT_RX "src=" MACSTR
-		" freq=%u type=%d", MAC2STR(src), freq, type);
+		" freq=%u type=%d rssi=%d",
+		MAC2STR(src), freq, type, ssi_signal);
 
 #ifdef CONFIG_DPP2
 	if (dpp_relay_rx_action(hapd->iface->interfaces->dpp,
--- a/src/ap/dpp_hostapd.h
+++ b/src/ap/dpp_hostapd.h
@@ -20,7 +20,8 @@
 int hostapd_dpp_listen(struct hostapd_data *hapd, const char *cmd);
 void hostapd_dpp_listen_stop(struct hostapd_data *hapd);
 void hostapd_dpp_rx_action(struct hostapd_data *hapd, const u8 *src,
-			   const u8 *buf, size_t len, unsigned int freq);
+			   const u8 *buf, size_t len, unsigned int freq,
+			   int ssi_signal);
 void hostapd_dpp_tx_status(struct hostapd_data *hapd, const u8 *dst,
 			   const u8 *data, size_t data_len, int ok);
 struct wpabuf *
--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -1845,7 +1845,8 @@
 		pos = mgmt->u.action.u.vs_public_action.oui;
 		end = drv_mgmt->frame + drv_mgmt->frame_len;
 		hostapd_dpp_rx_action(hapd, mgmt->sa, pos, end - pos,
-				      drv_mgmt->freq);
+				      drv_mgmt->freq,
+				      drv_mgmt->ssi_signal);
 		return;
 	}
 #endif /* CONFIG_DPP */
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -6797,7 +6797,7 @@
 
 static int handle_action(struct hostapd_data *hapd,
 			 const struct ieee80211_mgmt *mgmt, size_t len,
-			 unsigned int freq)
+			 unsigned int freq, int ssi_signal)
 {
 	struct sta_info *sta;
 	u8 *action __maybe_unused;
@@ -6906,7 +6906,7 @@
 			pos = mgmt->u.action.u.vs_public_action.oui;
 			end = ((const u8 *) mgmt) + len;
 			hostapd_dpp_rx_action(hapd, mgmt->sa, pos, end - pos,
-					      freq);
+					      freq, ssi_signal);
 			return 1;
 		}
 		if (len >= IEEE80211_HDRLEN + 2 &&
@@ -7159,7 +7159,7 @@
 		break;
 	case WLAN_FC_STYPE_ACTION:
 		wpa_printf(MSG_DEBUG, "mgmt::action");
-		ret = handle_action(hapd, mgmt, len, freq);
+		ret = handle_action(hapd, mgmt, len, freq, ssi_signal);
 		break;
 	default:
 		hostapd_logger(hapd, mgmt->sa, HOSTAPD_MODULE_IEEE80211,
--- a/wpa_supplicant/dpp_supplicant.c
+++ b/wpa_supplicant/dpp_supplicant.c
@@ -4255,7 +4255,7 @@
 
 void wpas_dpp_rx_action(struct wpa_supplicant *wpa_s, const u8 *src,
 			const u8 *buf, size_t len, unsigned int freq,
-			unsigned int freq_offset)
+			unsigned int freq_offset, int ssi_signal)
 {
 	u8 crypto_suite;
 	enum dpp_public_action_frame_type type;
@@ -4290,19 +4290,21 @@
 		wpa_printf(MSG_DEBUG, "DPP: Unsupported crypto suite %u",
 			   crypto_suite);
 		wpa_msg(wpa_s, MSG_INFO, DPP_EVENT_RX "src=" MACSTR
-			" freq=%u freq_offset=%d type=%d ignore=unsupported-crypto-suite",
-			MAC2STR(src), freq, freq_offset, type);
+			" freq=%u freq_offset=%d type=%d ignore=unsupported-crypto-suite"
+			" rssi=%d",
+			MAC2STR(src), freq, freq_offset, type, ssi_signal);
 		return;
 	}
 	wpa_hexdump(MSG_MSGDUMP, "DPP: Received message attributes", buf, len);
 	if (dpp_check_attrs(buf, len) < 0) {
 		wpa_msg(wpa_s, MSG_INFO, DPP_EVENT_RX "src=" MACSTR
-			" freq=%u freq_offset=%d type=%d ignore=invalid-attributes",
-			MAC2STR(src), freq, freq_offset, type);
+			" freq=%u freq_offset=%d type=%d ignore=invalid-attributes"
+			" rssi=%d",
+			MAC2STR(src), freq, freq_offset, type, ssi_signal);
 		return;
 	}
-	wpa_msg(wpa_s, MSG_INFO, DPP_EVENT_RX "src=" MACSTR " freq=%u freq_offset=%d type=%d",
-		MAC2STR(src), freq, freq_offset, type);
+	wpa_msg(wpa_s, MSG_INFO, DPP_EVENT_RX "src=" MACSTR " freq=%u freq_offset=%d type=%d rssi=%d",
+		MAC2STR(src), freq, freq_offset, type, ssi_signal);
 
 	switch (type) {
 	case DPP_PA_AUTHENTICATION_REQ:
--- a/wpa_supplicant/dpp_supplicant.h
+++ b/wpa_supplicant/dpp_supplicant.h
@@ -28,7 +28,7 @@
 void wpas_dpp_tx_wait_expire(struct wpa_supplicant *wpa_s);
 void wpas_dpp_rx_action(struct wpa_supplicant *wpa_s, const u8 *src,
 			const u8 *buf, size_t len, unsigned int freq,
-			unsigned int freq_offset);
+			unsigned int freq_offset, int ssi_signal);
 int wpas_dpp_configurator_sign(struct wpa_supplicant *wpa_s, const char *cmd);
 int wpas_dpp_pkex_add(struct wpa_supplicant *wpa_s, const char *cmd);
 int wpas_dpp_pkex_remove(struct wpa_supplicant *wpa_s, const char *id);
--- a/wpa_supplicant/events.c
+++ b/wpa_supplicant/events.c
@@ -5674,7 +5674,7 @@
 		payload++;
 		plen--;
 		wpas_dpp_rx_action(wpa_s, mgmt->sa, payload, plen, freq,
-				   freq_offset);
+				   freq_offset, rssi);
 		return;
 	}
 #endif /* CONFIG_DPP */
