diff --git a/package/network/utils/iwinfo/Makefile b/package/network/utils/iwinfo/Makefile
index c6a14956fd..f4d141f922 100644
--- a/package/network/utils/iwinfo/Makefile
+++ b/package/network/utils/iwinfo/Makefile
@@ -10,11 +10,11 @@ PKG_NAME:=libiwinfo
 PKG_RELEASE:=1
 
 PKG_SOURCE_PROTO:=git
-PKG_SOURCE_URL=$(PROJECT_GIT)/project/iwinfo.git
-PKG_SOURCE_DATE:=2023-07-01
-PKG_SOURCE_VERSION:=ca79f64154b107f192ec3c1ba631816cb8b07922
-PKG_MIRROR_HASH:=5eddf584a1c3ed5637162d6bfc573ed1ce3691fcb38bdd55bf9f1e11e82ccc46
-PKG_MAINTAINER:=Jo-Philipp Wich <jo@mein.io>
+PKG_SOURCE_URL:=https://github.com/MorseMicro/iwinfo.git
+PKG_SOURCE_DATE:=2024-10-18
+PKG_SOURCE_VERSION:=88de1400de808127891ff9780d09fe56526d504b
+PKG_HASH:=261be17a411c45d29075c2484e25859da5982ab2ecb8049651e79c276a23a319
+PKG_MAINTAINER:=Morse Micro <info@morsemicro.com>
 PKG_LICENSE:=GPL-2.0
 
 PKG_BUILD_FLAGS:=no-lto
