From 5209d9e0fcf03bc5fb3d3bfcfa227e0d9a86bf01 Mon Sep 17 00:00:00 2001
From: James Haggerty <james.haggerty@morsemicro.com>
Date: Fri, 11 Apr 2025 10:46:19 +1000
Subject: [PATCH 2/2] APP-4377 Tune down thresholds and fix led exit

---
 led.c | 12 ++++++------
 mac.c |  2 ++
 2 files changed, 8 insertions(+), 6 deletions(-)

diff --git a/led.c b/led.c
index f21a6ccc..3b19eec3 100644
--- a/led.c
+++ b/led.c
@@ -17,12 +17,12 @@ static const struct ieee80211_tpt_blink morse_blink[] = {
 	{ .throughput = 512 - 1, .blink_time = 260 },
 	{ .throughput = 1024 - 1, .blink_time = 220 },
 	{ .throughput = 2 * 1024 - 1, .blink_time = 190 },
-	{ .throughput = 3 * 1024 - 1, .blink_time = 170 },
-	{ .throughput = 6 * 1024 - 1, .blink_time = 150 },
-	{ .throughput = 12 * 1024 - 1, .blink_time = 130 },
-	{ .throughput = 18 * 1024 - 1, .blink_time = 110 },
-	{ .throughput = 24 * 1024 - 1, .blink_time = 80 },
-	{ .throughput = 30 * 1024 - 1, .blink_time = 50 },
+	{ .throughput = 4 * 1024 - 1, .blink_time = 170 },
+	{ .throughput = 7 * 1024 - 1, .blink_time = 150 },
+	{ .throughput = 11 * 1024 - 1, .blink_time = 130 },
+	{ .throughput = 16 * 1024 - 1, .blink_time = 110 },
+	{ .throughput = 22 * 1024 - 1, .blink_time = 80 },
+	{ .throughput = 29 * 1024 - 1, .blink_time = 50 },
 };
 
 static void morse_led_callback(struct led_classdev *led_cdev, enum led_brightness brightness)
diff --git a/mac.c b/mac.c
index cef1ff71..218c9e49 100644
--- a/mac.c
+++ b/mac.c
@@ -6953,6 +6953,7 @@ int morse_mac_register(struct morse *mors)
 		ret = ieee80211_register_hw(hw);
 	if (ret) {
 		MORSE_ERR(mors, "ieee80211_register_hw failed %d\n", ret);
+		morse_led_exit(mors);
 		goto err;
 	}
 
@@ -7029,6 +7030,7 @@ err_free_mon:
 #endif
 
 err_unregister_hw:
+	morse_led_exit(mors);
 	ieee80211_unregister_hw(hw);
 err:
 	return ret;
-- 
2.43.0

