Index: mesh11sd-1.2.0/linux_openwrt/mesh11sd/files/etc/init.d/mesh11sd
===================================================================
--- mesh11sd-1.2.0.orig/linux_openwrt/mesh11sd/files/etc/init.d/mesh11sd
+++ mesh11sd-1.2.0/linux_openwrt/mesh11sd/files/etc/init.d/mesh11sd
@@ -11,3 +11,12 @@
 	procd_close_instance
 }
 
+reload_service() {
+        # Restarts wireless interface to apply mesh config
+        wifi up
+}
+
+service_triggers() {
+        procd_add_reload_trigger "mesh11sd"
+}
+
