--- a/arch/arm/boot/dts/overlays/Makefile
+++ b/arch/arm/boot/dts/overlays/Makefile
@@ -143,7 +143,11 @@ dtbo-$(CONFIG_ARCH_BCM2835) += \
 	miniuart-bt.dtbo \
 	mipi-dbi-spi.dtbo \
 	mlx90640.dtbo \
+	mm_wlan.dtbo \
 	mmc.dtbo \
+	morse-ps.dtbo \
+	mm610x-spi.dtbo \
+	mm810x-spi.dtbo \
 	mpu6050.dtbo \
 	mz61581.dtbo \
 	ov2311.dtbo \
@@ -173,6 +177,7 @@ dtbo-$(CONFIG_ARCH_BCM2835) += \
 	pwm-ir-tx.dtbo \
 	qca7000.dtbo \
 	qca7000-uart0.dtbo \
+	ramoops.dtbo \
 	rotary-encoder.dtbo \
 	rpi-backlight.dtbo \
 	rpi-cirrus-wm5102.dtbo \
